pingcap/tidb
# *: fix sem dynamic privilege permission issues

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25314 close https://github.com/pingcap/tidb/issues/25312
close https://github.com/pingcap/tidb/issues/25350

Problem Summary:

This fixes 3 bugs discovered in SEM:
- The `RESTRICTED_CONNECTION_ADMIN` privilege did not work correctly.
- The commands `backup` and `restore` did not correctly prevent local filesystem access.
- `RESTRICTED_USER_ADMIN` users could not kill their own connection unless they also had `RESTRICTED_CONNECTION_ADMIN`. This is not correct - you should always be able to kill your own connections.

They are not strictly related, but all require cherry picking + a change to tidb-test, so it is convenient to fix them together.

### What is changed and how it works?

What's Changed:

One new dynamic privileges are added.
Bugs fixed in an unreleased version.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Adding kill tests in the mock session manager does not work correctly, and requires a complete tidb-server. This is the test scenario I tried:

```
func (s *testSuite3) TestSecurityEnhancedModeRestrictedConnectionAdmin(c *C) {
	tk := testkit.NewTestKit(c, s.store)
	tk.MustExec("CREATE USER ordinaryuser, normalkill, restricteduser, restrictedconnadmin ")
	tk.MustExec("GRANT SUPER, PROCESS, CONNECTION_ADMIN ON *.* to normalkill WITH GRANT OPTION")
	tk.MustExec("GRANT RESTRICTED_USER_ADMIN ON *.* to restricteduser WITH GRANT OPTION")
	// You must have both connection_admin and restricted_connection_admin to kill a restricted connection (limitation of privilege system not permitting OR conditions)
	// It usually makes sense to also have PROCESS so you can run SHOW PROCESSLIST before killing connections, but it's not strictly required.
	// Restricted_connection_admin means you can kill connections belonging to a RESTRICTED_USER_ADMIN, not that you are a restricted_user_admin
	tk.MustExec("GRANT CONNECTION_ADMIN, PROCESS, RESTRICTED_CONNECTION_ADMIN ON *.* to restrictedconnadmin WITH GRANT OPTION")

	sem.Enable()
	defer sem.Disable()

	ordinary, err := session.CreateSession4Test(s.store)
	c.Check(err, IsNil)
	defer ordinary.Close()
	//	sm := &mockSessionManager{}
	c.Assert(ordinary.Auth(&auth.UserIdentity{Username: "ordinaryuser", Hostname: "localhost"}, nil, nil), IsTrue)
	//	ordinary.SetSessionManager(sm)

	normalkill, err := session.CreateSession4Test(s.store)
	c.Check(err, IsNil)
	c.Assert(normalkill.Auth(&auth.UserIdentity{Username: "normalkill", Hostname: "localhost"}, nil, nil), IsTrue)
	//	normalkill.SetSessionManager(sm)

	go func() {
		_, err = ordinary.ExecuteInternal(context.Background(), "SELECT SLEEP(30)")
		c.Check(err, IsNil)
	}()

	// The normalkill user should be able to kill the ordinary connectionId because
	// they have connectionAdmin and the user does not have restrictedConnectionAdmin
	connectionId := ordinary.GetSessionVars().ConnectionID
	_, err = normalkill.ExecuteInternal(context.Background(), fmt.Sprintf("KILL TIDB %d", connectionId))
	c.Assert(err, IsNil)

	// Now create a sleeping query as a non-killable user
	restricteduser, err := session.CreateSession4Test(s.store)
	c.Check(err, IsNil)
	c.Assert(restricteduser.Auth(&auth.UserIdentity{Username: "restricteduser", Hostname: "localhost"}, nil, nil), IsTrue)
	//	restricteduser.SetSessionManager(sm)
	go func() {
		_, err = restricteduser.ExecuteInternal(context.Background(), "SELECT SLEEP(30)")
		c.Check(err, IsNil)
	}()

	// Try and kill a connection owned by restricteduser
	connectionId = restricteduser.GetSessionVars().ConnectionID
	_, err = normalkill.ExecuteInternal(context.Background(), fmt.Sprintf("KILL TIDB %d", connectionId))
	c.Assert(err, NotNil)

}
```

Unfortunately the last assertion always returns `Nil` in the tests, but returns `NotNil` in a manual test. I've tried a few different variations to try and get this working, but without luck.

Side effects

- Increased code complexity

### Release note 

- No release note
